spanner/session.go

```diff
@@ -633,6 +686,69 @@ func (p *sessionPool) recordStat(ctx context.Context, m *stats.Int64Measure, n i
 	recordStat(ctx, m, n)
 }
 
+func (p *sessionPool) getRatioOfSessionsInUseLocked() float64 {
+	maxSessions := p.MaxOpened
+	if maxSessions == 0 {
+		return 0
+	}
+	return float64(p.numInUse) / float64(maxSessions)
+}
+
+// gets sessions which are unexpectedly long-running.
+func (p *sessionPool) getLongRunningSessionsLocked() []*sessionHandle {
+	usedSessionsRatio := p.getRatioOfSessionsInUseLocked()
+	var longRunningSessions []*sessionHandle
+	if usedSessionsRatio > p.usedSessionsRatioThreshold {
+		element := p.trackedSessionHandles.Front()
+		for element != nil {
+			sh := element.Value.(*sessionHandle)
+			sh.mu.Lock()
+			diff := time.Now().Sub(sh.checkoutTime)
+			if !sh.isLongRunningTransaction && diff.Seconds() >= p.idleTimeThreshold.Seconds() {
+				if p.LogInactiveTransactions && !sh.isSessionLeakLogged {
+					if sh.stack != nil {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running due to possible session leak for goroutine: \n%s", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339), sh.stack)
+					} else {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running due to possible session leak for goroutine: \nStack trace not available", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339))
+					}
+					sh.isSessionLeakLogged = true
+				}
+				if p.CloseInactiveTransactions && !sh.isSessionLeakLogged {
+					longRunningSessions = append(longRunningSessions, sh)
+					if sh.stack != nil {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running and will be removed due to possible session leak for goroutine: \n%s", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339), sh.stack)
+					} else {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running and will be removed due to possible session leak for goroutine: \nStack trace not available", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339))
+					}
+					sh.isSessionLeakLogged = true
+				}
+			}
+			sh.mu.Unlock()
+			element = element.Next()
+		}
+	}
+	return longRunningSessions
+}
+
+// removes or logs sessions that are unexpectedly long-running.
+func (p *sessionPool) removeLongRunningSessions() {
+	p.mu.Lock()
+	longRunningSessions := p.getLongRunningSessionsLocked()
+	p.mu.Unlock()
+
+	// destroy long-running sessions
+	if p.CloseInactiveTransactions {
+		for _, sh := range longRunningSessions {
```

Comment:
We can't do this inline (at time of detection) and avoid re-iterating this list? Is there other reasons why you are preferring a two-step approach where you are first fetching the long running sessions and then re-iterating the list and destroying?

Reply:
The below are the reasons why it cannot be done inline (destroying at the time of detection)
1. Locking is one reason. The existing `sh.destroy()` method takes a lock on session pool internally. Similarly, in `getLongRunningSessionsLocked` method, to iterate `p.trackedSessionHandles` (linked list that contains sessions in use) we need to take a lock on session pool. This means if we want to call destroy during iteration, then we need to remove the lock on session pool, then call the destroy method and then wait for the next lock to be acquired on the session pool for moving ahead with the iteration.
2. Also, looking at the `getLongRunningSessionsLocked` method, the  `trackedSessionHandles` is a **linked list**, which contains a list of all the sessions that are currently in use. When iterating over a linked list using a loop, we cannot directly delete values from the list during the iteration which can result in unexpected behaviour or runtime errors.

Thus a two-step approach is the preferable and safer operation.